### PR TITLE
Add detailed style guidance for writing about YAML

### PR DESCRIPTION
**Pull request by ddbeck**
YAML is a little bit of a nightmare: You studied! You should know this! But you're late for the exam! 😱

If there's one thing I want to communicate it is this: don't talk about YAML qua YAML if you can avoid it. As much as possible, talk about the things it does for the reader—creating pipelines, modifying behavior—not the act of authoring YAML.

To come up with this stuff, I looked at what the YAML specification uses and what the existing docs do and what some other documentation that mentions YAML does. This is what I came up with, in attempt to (mostly) capture the most common usage in the Buildkite docs already and make some useful distinctions, where there are conflicts in the existing text (of which there are many).

### CODE REVIEW

File: styleguide/yaml.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Talking about YAML
```
Comment (karensawrey) [inline styleguide/yaml.md:1]:
```suggestion
# Talking about YAML

This document is a part of the larger [style guide](./STYLE.md) and it covers the usage of YAML in the Buildkite documentation. Further expansion of this guide is highly welcomed.
```

**Review by karensawrey — CHANGES_REQUESTED** [review]
Great addition! Needs a few tiny tweaks before we merge it.

**Review by karensawrey — APPROVED** [review]
Looks good!

File: styleguide/yaml.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+❌ Add the `skip` attribute to the command step, then on a new line…
+
+## Key and value
+
+Use *key* to refer only to an attribute's identifier and *value* to refer only to the attribute's value.
+Remember to use code formatting for literal keys or values.
+
+✅ Add the `skip` key to the command step, then on a new line…<br>
+✅ Set the value to `true`, `false`, or a string…<br>
+❌ Add the skip key to the command step, then on a new line…
```
[the remark this answers is not in the document]
Comment (ddbeck) [reply styleguide/yaml.md:52]: Oh wait, I forgot to address this!

So, I kinda-sorta covered this in the earlier maps section, like so:

> If you must describe a map within a map or other structure, qualify the relationship with *of* or use the term *nested map*. Do not use the terms *block*, *level*, or the prefix *sub*.

My instinct here was to discourage using "levels" to describe the relationship between nested maps. The vertical level metaphor works pretty well with the block-style notation, but not so much with the flow-style notation (e.g., given `{ a: { b: false } }`, `b` lacks any visual leveliness, for lack of a better word). And it falls apart quickly when you involve arrays (is a map in an array in a map second or third level?).

That said, I could write an exception for the top-level map. However, I think it's avoidable entirely (calling back to the "refer to what the YAML represents" directive at the outset), so maybe we could see whether we can get away without it?

[the remark this answers is not in the document]
Comment (karensawrey) [reply styleguide/yaml.md:52]: I must confess that the more I read the proposed improvement regarding the block vs flow-style notation, the more confusing it gets.

[the remark this answers is not in the document]
Comment (ddbeck) [reply styleguide/yaml.md:52]: It _is_ confusing. I spent a bit of time reading the YAML spec, so I've got a long head starting thinking about how to talk about YAML. But fundamentally, the problem is this: YAML has lots of ways to record the same data (e.g., inline versus block, quoted versus unquoted, etc.) and it's very easy to overdetermine what we're asking for.

Consider this YAML source:

```yaml
meal:
  type: sandwich
  size: small
  vegan: false
```

It's directly equivalent to this YAML source:

```yaml
{ meal: { type: "sandwich", size: "small", vegan: FALSE } }
```

`meal` is a "top-level" map in both, but in the latter case the top-level map is metaphorical, as there's no visual vertical hierarchy depicted by the source. What we really care about is that readers use the right structure and values, not the specific notation that we prescribe.

[the remark this answers is not in the document]
Comment (karensawrey) [reply styleguide/yaml.md:52]: After giving this more thought... I think we'd be better off merging the addition to the style guide vs trying to come up with a perfect way to address map/levels and never merging it ;) So please finalize the PR as you see fit in that regard. 🙏  I'm very much OK with the rest of the PR I've already reviewed. 👍

**Comment by plaindocs** [comment]
I'll dig into this when I get back, looks like thorough work.

File: styleguide/yaml.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (plaindocs) [reply styleguide/yaml.md:52]: I've **finally** made it here.  

I think we can focus exclusively on the more human-readable line-break and whitespaced YAML, if you'll permit

:heavy_check_mark:  

```yaml
meal: 
  type: sandwich 
  size: small 
  vegan: false
```

❌

```yaml
{ meal: { type: "sandwich", size: "small", vegan: FALSE } }
```

And I think we can **probably** ignore the indenting and trust the linter (which we don't yet have for codeblocks) to sort that out.

**Review by plaindocs — APPROVED** [review]
This is great. 

How about we add a _Buildkitey_ example of Map, Attribute, Key and value and Array, and merge this in.

Thanks for your patience @ddbeck

File: styleguide/yaml.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ddbeck) [reply styleguide/yaml.md:83]: Cool, I've added this as a general guideline, closer to the top. I also added a bit more structure to the whole thing, so it's easier to ignore the bits that won't apply in the usual course of events.

**Comment by plaindocs** [comment]
Sorry @ddbeck one more change (addition now), which I did not communicate well.

I'd love to see a YAML example in each of the H3 headings inside `## Terms`, ie, this is what a YAML array looks like, this is what a YAML map looks like. And I'd love to see that using examples from the docs or similar, not sandwich types. (although I love Sandwiches)

Thanks for tackling this guide!

**Pull request merged**